lms/djangoapps/courseware/masquerade.py

```diff
@@ -116,3 +164,79 @@ def get_masquerading_group_info(user, course_key):
     if not course_masquerade:
         return None, None
     return course_masquerade.group_id, course_masquerade.user_partition_id
+
+
+# Sentinel object to mark deleted objects in the session cache
+_DELETED_SENTINEL = object()
+
+
+class MasqueradingKeyValueStore(KeyValueStore):
+    """
+    A `KeyValueStore` to avoid affecting the user state when masquerading.
+
+    This `KeyValueStore` wraps an underlying `KeyValueStore`.  Reads are forwarded to the underlying
+    store, but writes go to a Django session (or other dictionary-like object).
+    """
+    def __init__(self, kvs, session):
+        """
+        Arguments:
+          kvs: The KeyValueStore to wrap.
+          session: The Django session used to store temporary data in.
+        """
+        self.kvs = kvs
+        self.session = session
+        self.session_data = session.setdefault(MASQUERADE_DATA_KEY, {})
+
+    def _serialize_key(self, key):
+        """
+        Convert the key of Type KeyValueStore.Key to a string.
+
+        Keys are not JSON-serializable, so we can't use them as keys for the Django session.
+        The implementation is taken from cms/djangoapps/contentstore/views/session_kv_store.py.
+        """
+        return repr(tuple(key))
+
+    def get(self, key):
+        try:
+            value = self.session_data[self._serialize_key(key)]
+        except KeyError:
+            return self.kvs.get(key)
+        else:
+            if value is _DELETED_SENTINEL:
+                raise KeyError(key.field_name)
+            return value
+
+    def set(self, key, value):
+        self.session_data[self._serialize_key(key)] = value
+        self.session.modified = True
+
+    def delete(self, key):
+        # We can't simply delete the key from the session, since it might still exist in the kvs,
+        # which we are not allowed to modify, so we mark it as deleted by setting it to
+        # _DELETED_SENTINEL in the session.
+        self.set(key, _DELETED_SENTINEL)
+
+    def has(self, key):
+        try:
+            value = self.session_data[self._serialize_key(key)]
+        except KeyError:
+            return self.kvs.has(key)
+        else:
+            return value != _DELETED_SENTINEL
+
+
+def filter_displayed_blocks(block, unused_view, frag, unused_context):
+    """
+    A wrapper to only show XBlocks that set `show_in_read_only_mode` when masquerading as a specific user.
+
+    We don't want to modify the state of the user we are masquerading as, so we can't show XBlocks
+    that store information outside of the XBlock fields API.
+    """
+    if getattr(block, 'show_in_read_only_mode', False):
+        return frag
+    return Fragment(
+        u'This type of block cannot be shown while viewing the course as a specific student. '
```

Comment:
@catong "Problem" might be too limited in scope.  I think the term used in the edX documentation term is "component": http://edx.readthedocs.org/en/latest/glossary.html#c